feat(agents): NPC tick with ElizaOS integration and game awareness

SYMBaiEX opened the pull request:
## Summary

- **Separate NPC tick cron**: NPCs now run on their own dedicated `/api/cron/npc-tick` endpoint, decoupled from user agent processing
- **ElizaOS game context**: NPCs receive arc awareness (insider/deceiver status, narrative phases) via new `npcGameContextProvider`
- **Anti-slop quality rules**: Shared `NPC_POST_QUALITY_RULES` prevent robotic/technical content in NPC posts

## Type

- [x] Feature
- [ ] Bug fix
- [x] Refactor / cleanup (no behavior change)
- [ ] Performance
- [ ] Infra / ops
- [ ] Docs
- [ ] Contracts / on-chain
- [ ] Other: …

## Context / Links

- Fixes feed quality issues where NPCs were generating overly technical, market-analyst-style content
- NPCs now use the same `AutonomousCoordinator` as user agents but with `isNpc=true` flag for game awareness injection

## Scope (keep it focused)

- [x] This PR is focused on a single change/theme (not a catch-all)
- [x] Drive-by refactors are excluded or split into a separate PR
- [x] Non-goals / follow-ups are listed below (with links)

**Areas touched**
- [ ] Web UI (`apps/web`)
- [x] Web API routes / SSE / A2A (`apps/web`)
- [ ] CLI (`apps/cli`)
- [ ] Docs site (`apps/docs`) / vendor docs (`docs/vendors/*`)
- [x] Domain / game engine (`packages/engine`, `packages/core/*`)
- [x] Agents / runtime / A2A / MCP (`packages/agents`, `packages/a2a`, `packages/mcp`)
- [ ] API infra (`packages/api`)
- [ ] DB (`packages/db`)
- [ ] Contracts / on-chain (`packages/contracts`)
- [ ] Shared types/utils (`packages/shared`)
- [ ] Tests (`packages/testing`)
- [ ] Other: …

## Changes

- **New**: `apps/web/src/app/api/cron/npc-tick/route.ts` - Dedicated NPC cron job
- **New**: `packages/agents/src/plugins/babylon/providers/npc-game-context.ts` - ElizaOS provider for arc awareness
- **New**: `packages/engine/src/prompts/shared-sections.ts` - Added `NPC_POST_QUALITY_RULES`
- **Modified**: `packages/agents/src/autonomous/MultiStepExecutor.ts` - Injects NPC context when `isNpc=true`
- **Modified**: `packages/agents/src/autonomous/templates/multi-step-decision.ts` - Includes quality rules for NPCs
- **Modified**: `packages/engine/src/game-tick.ts` - Removed NPC post generation (now in npc-tick)
- **Modified**: `apps/web/src/app/api/cron/agent-tick/route.ts` - Excludes NPCs (handled by npc-tick)
- **Modified**: `vercel.json` - Added npc-tick cron schedule and function config

## Review guide

**Start here**
- `apps/web/src/app/api/cron/npc-tick/route.ts` - Main entry point for NPC processing
- `packages/agents/src/plugins/babylon/providers/npc-game-context.ts` - How NPCs get game awareness

**Risk**
- [ ] Low
- [x] Medium
- [ ] High

**Notes for reviewers**
- NPCs rotate through 20 per minute to ensure all get processed over time
- Game context is injected as natural language "intuitions" (no raw game mechanics exposed)
- Existing `NPCBootstrapService` registers NPCs at server startup via `instrumentation.ts`

## Test plan

**Commands run**
- [ ] `bun run check` (Biome format)
- [x] `bun run typecheck`
- [x] `bun run lint`
- [ ] `bun run build`
- [x] `bun run test` (unit + integration) - NPC system validation tests pass
- [ ] `bun run test:e2e` (if critical flows changed)
- [ ] `bun run contracts:test` (if contracts changed)
- [ ] `bun run db:check` (if DB schema/queries changed)

**Manual verification**
1. NPC tick endpoint returns valid JSON response
2. Agent tick excludes NPCs (only USER_CONTROLLED agents processed)
3. Integration tests pass for agent-autonomous-tick

## Ops / Migration / Deployment

- [ ] No deploy impact
- [ ] Requires env var updates (listed below + `.env.example` updated)
- [ ] Requires DB migration (`bun run db:migrate`) / backfill / seed
- [x] Changes cron schedule or endpoints (`vercel.json`, `CRON_SECRET`, etc.)
- [ ] Rollout behind a flag / gradual rollout

**Env vars (added/changed/removed)**
- None

**DB / data migration**
- None

**Rollout / rollback plan**
- Rollout: Deploy to staging, verify NPC tick runs successfully every minute
- Rollback: Revert commit, NPCs will stop posting until game-tick NPC generation is restored

## Breaking changes

- [x] None
- [ ] Yes (describe + migration guide below)

## Security / privacy

- [x] No security impact
- [ ] Needs extra review (describe)

## Screenshots / recordings (UI)

- N/A (backend only)

## Checklist (author)
- [x] Self-review done (diff + critical paths)
- [x] Base branch is correct (`staging` by default)
- [x] Handlers remain thin and portable (validate → service → map errors)
- [x] Domain logic stays in packages (no Next/React/Elysia coupling in core)
- [ ] `.env.example` updated (if env changed) and variables documented above
- [ ] Docs updated (and `bun run docs:generate` if needed)
- [ ] Tests added/updated for behavior changes (or rationale provided)
- [ ] Dependency changes are intentional (`bun.lock` updated)
- [ ] Code owners requested (auto via CODEOWNERS or manual)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * NPC-specific scheduled tick for autonomous NPC activity.
  * Desktop leaderboard widget with tab toggle and per-user detail sidebar.
  * Article cover image generation with retry; posts and articles now include imageUrl.
  * Server-side caching for the leaderboard API.

* **Refactor**
  * Main tick refocused on organization content; NPC processing moved to a separate workflow.
  * NPC-aware context and NPC-specific prompt/rule support added.

* **Tweaks**
  * UI refinements across notifications, back buttons, markets, and wallet displays.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SYMBaiEX committed:
feat(api): add npc-tick endpoint and refactor agent-tick logic

- Introduced a new /api/cron/npc-tick endpoint to handle NPC agent processing separately from user-controlled agents.
- Updated agent-tick logic to focus solely on USER_CONTROLLED agents, improving clarity and maintainability.
- Enhanced logging to differentiate between user agents and NPCs, ensuring accurate tracking of agent activity.
- Adjusted the MultiStepExecutor to retrieve NPC game context, enriching NPC decision-making with relevant world events.
- Updated various components and prompts to incorporate NPC-specific quality rules for more authentic social media interactions.

SYMBaiEX committed:
refactor: improve code formatting and readability across multiple files

- Adjusted line breaks and indentation for better readability in route handlers and components.
- Enhanced logging statements for clarity in the game-tick and NPC processing logic.
- Refactored conditional rendering in the leaderboard page for improved structure.
- Cleaned up import statements in the LeaderboardWidgetSidebar for consistency.

SYMBaiEX committed:
refactor: clean up import statements and improve code formatting

- Removed unnecessary imports and adjusted import order for consistency across multiple files.
- Enhanced code readability by improving line breaks and indentation in various route handlers and components.
- Added logging for better clarity in NPC processing and agent management.
- Updated the MultiStepExecutor to use a default balance for NPCs without an actorState record, ensuring robust handling of missing data.

SYMBaiEX committed:
fix: remove unnecessary whitespace in LeaderboardWidgetSidebar and npc-game-context

- Cleaned up whitespace in LeaderboardWidgetSidebar to enhance code readability.
- Removed an extra newline in npc-game-context for improved formatting consistency.

SYMBaiEX committed:
refactor: update agent configuration and improve command handling

- Changed the RPC URL environment variable in the agent configuration from BASE_SEPOLIA_RPC_URL to AGENT0_RPC_URL for clarity.
- Refactored the command handling in the model command to ensure database closure is managed properly using a try-finally block, enhancing error handling and resource management.
- Improved caching logic in the leaderboard API to optimize performance and response times by implementing cache checks and storing responses effectively.
- Enhanced styling in the PerpDetailPage and PredictionDetailPage components for better user experience and consistency.
- Updated the NotificationsPage layout for improved accessibility and visual structure, ensuring a more user-friendly interface.

SYMBaiEX committed:
refactor: improve layout formatting in NotificationsPage

- Adjusted the formatting of the PageContainer component in NotificationsPage for better readability and consistency.
- Enhanced the overall structure of the component to ensure a more user-friendly interface.

SYMBaiEX committed:
refactor: enhance command handling and caching logic

- Added 'upload-dataset' to the list of commands that don't require a database in the model command, improving command handling.
- Simplified the caching logic in the leaderboard API by using a fallback for CACHE_TTL_MS, ensuring consistent behavior and performance.

SYMBaiEX committed:
feat: add imageUrl field to posts and enhance article display

- Introduced imageUrl field in the posts schema and database service to support article cover images.
- Updated API routes to include imageUrl in post responses.
- Enhanced ArticlePage and ArticleCard components to display article cover images conditionally.
- Improved styling for article images in both mobile and desktop views for better user experience.
- Integrated image generation logic in the game tick process to automatically generate article cover images.

SYMBaiEX committed:
refactor: clean up import statements and enhance code organization

- Reformatted import statements in game-tick and article-image-service files for improved readability and consistency.
- Removed duplicate import of logger in article-image-service, streamlining the code structure.
- Adjusted import order in post-generation-helpers to maintain a clear and organized structure.

SYMBaiEX committed:
feat: enhance npc-tick logic with configurable parameters and circuit breaker

- Introduced environment variables for NPC tick batch size and maximum consecutive errors to improve configurability.
- Implemented caching for game state retrieval to reduce database load.
- Enhanced NPC processing with a circuit breaker mechanism to prevent cascading failures during execution.
- Improved wrap-around logic for NPC rotation to ensure robust handling regardless of array size.
- Updated metrics recording to reflect circuit breaker status and processed actions accurately.

SYMBaiEX committed:
refactor: clean up import statements in npc-tick route

- Removed duplicate import of Game type for improved code organization.
- Adjusted import order for better readability and consistency in the npc-tick route file.

slkzgm reviewed (commented):
I have a few questions before approval:

1) NPC game context: `generateWorldContext(false)` returns no headlines and the code doesn't surface `worldFacts.general`, so the "WHAT'S HAPPENING" section looks empty. Was that intentional? If you want that section populated, should this use `generatePromptContext()` or include `worldFacts.general`?

2) Article image generation: the code/comment says non-blocking, but `fal.subscribe(...)` will throw and fail the tick. Is the intent to fail fast if image generation fails, or should it be best-effort (swallow errors and continue)?

Optional: I don't see `NPC_TICK_BATCH_SIZE`, `NPC_TICK_MAX_ERRORS`, `AGENT_LOCK_DURATION_MS`, or `LEADERBOARD_CACHE_MS` documented in `.env.example`. Was that intentional? Is there default values set?

SYMBaiEX committed:
feat: add NPC tick configuration and improve article image generation error handling

- Introduced new environment variables for NPC tick configuration, including batch size, maximum errors, agent lock duration, and leaderboard cache TTL.
- Updated NPC game context to utilize general world facts for improved context generation.
- Enhanced article image generation function to log errors without blocking execution, ensuring robustness in the tick process.

SYMBaiEX committed:
fix: enhance NPC tick and MultiStepExecutor logic for improved robustness

- Updated NPC tick logic to include seconds-based entropy, preventing race conditions when multiple jobs fire within the same minute.
- Improved error handling in MultiStepExecutor by allowing graceful degradation if NPC game context fetch fails, ensuring continued operation without it.
- Added comments for clarity on the default outcome in the NPC game context, highlighting the need for future database integration.

SYMBaiEX committed:
refactor: simplify NPC tick calculation for improved clarity

- Adjusted the NPC tick calculation to enhance readability by removing unnecessary parentheses.
- Ensured the logic remains intact while improving the overall code organization in the npc-tick route file.

slkzgm commented on packages/engine/src/services/article-image-service.ts:
```diff
@@ -0,0 +1,181 @@
+  );
+
+  let result: FalResponse;
+  try {
```
Note: this introduces a try/catch. Repo guidelines (`.cursorrules`) say to avoid try/catch / fail fast. If that rule is still strict, consider documenting this as an exception or moving the error handling up a layer.

slkzgm reviewed (dismissed):
Summary: The NPC tick split + game context wiring look solid, and the env/docs updates line up with the new behavior. No blocking issues found.

Testing: not run (review only).

SYMBaiEX committed:
fix: improve error handling in article image generation for non-critical failures

- Updated the article image generation function to log failures without blocking execution, ensuring the game tick continues running smoothly.
- Added comments to clarify the intentional use of try-catch for external API interactions, emphasizing the non-critical nature of image generation failures.
